lighthouse-core/report/html/renderer/details-renderer.js

```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
    * @return {LH.Audit.Details.OpportunityColumnHeading}
    */
   _getCanonicalizedHeading(heading) {
-    let subRows;
-    if (heading.subRows) {
+    let subHeading;
+    if (heading.subHeading) {
       // @ts-ignore: It's ok that there is no text.
-      subRows = this._getCanonicalizedHeading(heading.subRows);
-      if (!subRows.key) {
+      subHeading = this._getCanonicalizedHeading(heading.subHeading);
+      if (!subHeading.key) {
         // eslint-disable-next-line no-console
         console.warn('key should not be null');
       }
-      subRows = {...subRows, key: subRows.key || ''};
+      subHeading = {...subHeading, key: subHeading.key || ''};
     }
 
     return {
       key: heading.key,
       valueType: heading.itemType,
-      subRows,
+      subHeading,
       label: heading.text,
       displayUnit: heading.displayUnit,
       granularity: heading.granularity,
     };
   }
 
   /**
-   * @param {LH.Audit.Details.ItemValue[]} values
-   * @param {LH.Audit.Details.OpportunityColumnHeading} heading
-   * @return {Element}
+   * @param {LH.Audit.Details.OpportunityItem | LH.Audit.Details.TableItem} item
+   * @param {(LH.Audit.Details.OpportunityColumnHeading | null)[]} headings
    */
-  _renderSubRows(values, heading) {
-    const subRowsElement = this._dom.createElement('div', 'lh-sub-rows');
-    for (const childValue of values) {
-      const subRowElement = this._renderTableValue(childValue, heading);
-      if (!subRowElement) continue;
-      subRowElement.classList.add('lh-sub-row');
-      subRowsElement.appendChild(subRowElement);
+  _renderTableRow(item, headings) {
+    const rowElem = this._dom.createElement('tr');
+
+    for (const heading of headings) {
+      const value = heading && heading.key && item[heading.key];
+
+      let valueElement;
+      if (heading && heading.key !== null && value !== undefined && value !== null) {
+        valueElement = this._renderTableValue(value, heading);
+      }
+
+      if (heading && valueElement) {
+        const classes = `lh-table-column--${heading.valueType}`;
+        this._dom.createChildOf(rowElem, 'td', classes).appendChild(valueElement);
+      } else {
+        // Empty cell is rendered for a column if:
+        // - the pair is null
+        // - the heading key is null
+        // - the value is undefined/null
+        this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
+      }
     }
-    return subRowsElement;
+
+    return rowElem;
+  }
+
+  /**
+   * Renders one or more rows from a details table item.
+   * @param {LH.Audit.Details.OpportunityItem | LH.Audit.Details.TableItem} item
+   * @param {LH.Audit.Details.OpportunityColumnHeading[]} headings
+   */
+  _renderTableItemRows(item, headings) {
```

Comment:
can we come up with some naming options for `_renderTableRow` and `_renderTableItemRows`? :)

I assume `_renderTableItemRows` is saying a table item can be a row or rows, depending on if it had subrows in it? But `_renderTableItem` also sounds like a singular thing (like `_renderTableValue`), not a row or rows...

`_renderTableRowsFromItem`?

Reply:
`_renderTableRowsFromItem` sgtm!